src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -274,12 +339,167 @@ public void WriteRawValue(string rawValue)
+        /// in an array.
+        /// This method should not be called by <see cref="WriteRawValue(string)"/>.
+        /// </summary>
+        private void WriteSeparatorIfNecessary()
+        {
+            if (this.shouldWriteSeparator)
+            {
+                this.WriteItemSeparator();
+            }
+
+            // reset state since now we're back in sync with Utf8JsonWriter.
+            this.shouldWriteSeparator = false;
+            this.isWritingFirstElementInArray = false;
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+        }
+
+        /// <summary>
+        /// Manually writes an item separator ',' to the output, bypassing the Utf8JsonWriter.
+        /// </summary>
+        private void WriteItemSeparator()
+        {
+            this.Flush();
+            this.writeStream.WriteByte(itemSeparator.Span[0]);
+        }
+
+        /// <summary>
+        /// Asynchronously manually writes an item separator ','
+        /// to the output if necessary, bypassing the Utf8JsonWriter.
+        /// Should be used before property names in an object or values
+        /// in an array.
+        /// This method should not be called by <see cref="WriteRawValue(string)"/>.
+        /// </summary>
+        /// <returns>Task representing the asynchronous operation.</returns>
+        private async ValueTask WriteSeparatorIfNecessaryAsync()
+        {
+            if (this.shouldWriteSeparator)
+            {
+                await this.WriteItemSeparatorAsync().ConfigureAwait(false);
+            }
+
+            // reset state since now we're back in sync with Utf8JsonWriter.
+            this.shouldWriteSeparator = false;
+            this.isWritingFirstElementInArray = false;
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+        }
+
+        /// <summary>
+        /// Asynchronously manually writes an item separator ',' to the output, bypassing the Utf8JsonWriter.
+        /// </summary>
+        /// <returns>Task respresenting the asynchronous operation.</returns>
+        private async ValueTask WriteItemSeparatorAsync()
+        {
+            await this.FlushAsync().ConfigureAwait(false);
+            await this.writeStream.WriteAsync(itemSeparator).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Mark that the writer has transitioned to write an object.
+        /// </summary>
+        private void EnterObjectScope()
+        {
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+            this.isWritingFirstElementInArray = false;
+            this.bitStack.PushTrue();
+        }
+
+        /// <summary>
+        /// Mark that the writer has transitioned to write an array.
+        /// </summary>
+        private void EnterArrayScope()
+        {
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+            this.isWritingFirstElementInArray = true;
+            this.bitStack.PushFalse();

```

Comment:
So for objectscope we push true to the bitstack then for array scope we push false?

Reply:
Yes, that's exactly what it means.